### PR TITLE
`in` uses `isequal` (#9381)

### LINKED ISSUE #9381: should `in` use `isequal` to test for containment?
This is how Dicts work, so there's a case to me made. Came up in this discussion:

https://groups.google.com/forum/#!topic/julia-users/XZDm57yHc5M

Motivating examples:

``` jl
julia> NaN in [1,1.5,2,NaN]
false

julia> NaN in Set([1,1.5,2,NaN])
true

julia> NaN in keys(Dict(1=>"one", 1.5=>"one and a half", 2=>"two", NaN=>"not a number"))
true
```

It seems like in for arrays is the one that's out of line here.

### PR DESCRIPTION
**Pull request by JackDevine**
I changed the docstring quite a bit, I think that it is now a good description of the current behavior. I also added the test in the first comment of #9381.

Edit: fix #9381

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
There are quite a few more methods of `in` – did you check all of them?

**Comment by JackDevine** [comment]
Thanks for pointing that out, I was just about to ask about that. For example, in associative.jl

```julia
function in(p::Pair, a::Associative, valcmp=(==))
    v = get(a,p[1],secret_table_token)
    if v !== secret_table_token
        valcmp(v, p[2]) && return true
    end
    return false
end
```
Do you want `valcmp` to become `isequal` by default? My understanding was that we are happy with how `in` works for dictionaries and sets. Also, would
```julia
in(x::Char, y::Char) = x == y
```
goes to
```julia
in(x::Char, y::Char) = isequal(x, y)
```
change anything?

When I look into this more, there are definitely a few that I need to change e.g. `in(x::Number, y::Number)`. Should I only make changes where they might make a difference or should I just be greedy and make lots of changes?

**Comment by StefanKarpinski** [comment]
Yes, I would change all of these. In some cases `==` behaves the same as `isequal` but if someone were to override or change the behavior of `isequal` but not `==` then we want this to be consistent everywhere.

File: NEWS.md [hunk 1]
```diff
@@ -272,6 +272,8 @@ This section lists changes that do not have deprecation warnings.
 Library improvements
 --------------------
 
+  * `in` now uses `isequal` to test for containment ([#9381]).
```
Comment (ararslan) [inline NEWS.md]: This should probably be listed under breaking changes instead.

Comment (JackDevine) [reply NEWS.md]: Good point, fixed.

**Comment by andyferris** [comment]
I also wonder if `==` for a container should be `isequal` for all the elements. For example, `[NaN] == [NaN]` - well, they are the same container, no?. Like `NaN`, this is also important for `null`/`missing`.

**Comment by JackDevine** [comment]
Is that idea relevant to this PR? I don't think that it is inconsistent to have `in` use `isequal` while still having `[NaN] !== [NaN]`.

**Comment by andyferris** [comment]
> Is that idea relevant to this PR?

Not in the sense that this PR needs modifying. :)

However, I think there is a fair amount of relevance to #9381 (actually I should have probably put my comment there instead, sorry!) in that I would say that *containers* would be equal when they *contain* (i.e. `element in container`) the same elements. Obviously `NaN in [NaN]` is relevant to this PR, but we also have `!([NaN] == [NaN])` but `Set(NaN) == Set(NaN)` which is *exactly* the kind of silliness that is being addressed here.

File: base/associative.jl [hunk 2]
```diff
@@ -67,7 +67,7 @@ function next(v::ValueIterator, state)
     n[1][2], n[2]
 end
 
-in(k, v::KeyIterator) = get(v.dict, k, secret_table_token) !== secret_table_token
+in(k, v::KeyIterator) = !isequal(get(v.dict, k, secret_table_token), secret_table_token)
```
Comment (JeffBezanson) [inline base/associative.jl]: This is not quite right. This is using `!==` (which is the opposite of `===`, not `==`) to see if exactly the `secret_table_token` object was returned.

**Review by JeffBezanson — CHANGES_REQUESTED** [review]

File: base/env.jl [hunk 3]
```diff
@@ -4,7 +4,7 @@ if Sys.iswindows()
     const ERROR_ENVVAR_NOT_FOUND = UInt32(203)
 
     _getenvlen(var::Vector{UInt16}) = ccall(:GetEnvironmentVariableW,stdcall,UInt32,(Ptr{UInt16},Ptr{UInt16},UInt32),var,C_NULL,0)
-    _hasenv(s::Vector{UInt16}) = _getenvlen(s) != 0 || Libc.GetLastError() != ERROR_ENVVAR_NOT_FOUND
+    _hasenv(s::Vector{UInt16}) = !isequal(_getenvlen(s), 0) || !isequal(Libc.GetLastError(), ERROR_ENVVAR_NOT_FOUND)
```
Comment (JeffBezanson) [inline base/env.jl]: These don't need to change.

File: base/env.jl [hunk 4]
```diff
@@ -38,7 +38,7 @@ if Sys.iswindows()
     end
 else # !windows
     _getenv(var::AbstractString) = ccall(:getenv, Cstring, (Cstring,), var)
-    _hasenv(s::AbstractString) = _getenv(s) != C_NULL
+    _hasenv(s::AbstractString) = !isequal(_getenv(s), C_NULL)
```
Comment (JeffBezanson) [inline base/env.jl]: Also this.

File: base/show.jl [hunk 5]
```diff
@@ -87,7 +87,7 @@ pipe_writer(io::IOContext) = io.io
 lock(io::IOContext) = lock(io.io)
 unlock(io::IOContext) = unlock(io.io)
 
-in(key_value::Pair, io::IOContext) = in(key_value, io.dict, ===)
+in(key_value::Pair, io::IOContext) = in(key_value, io.dict, isequal)
```
Comment (JeffBezanson) [inline base/show.jl]: This should not change since the caller was specifically requesting `===`.

File: base/strings/search.jl [hunk 6]
```diff
@@ -44,7 +44,7 @@ function search(s::AbstractString, c::Chars, i::Integer)
 end
 search(s::AbstractString, c::Chars) = search(s,c,start(s))
 
-in(c::Char, s::AbstractString) = (search(s,c)!=0)
+in(c::Char, s::AbstractString) = !isequal(search(s,c), 0)
```
Comment (JeffBezanson) [inline base/strings/search.jl]: This should also not change, since it isn't comparing `c`; it's testing the result of a different function.

**Comment by JackDevine** [comment]
Thanks @JeffBezanson , I may have gone a bit overboard with some of those changes. I have changed back to the old style in all of the places that you asked.

**Comment by JackDevine** [comment]
Closing due to comment in [#9381](https://github.com/JuliaLang/julia/issues/9381#issuecomment-355110628)

**Comment by ararslan** [comment]
Thank you regardless for your efforts here, @JackDevine!

**Pull request closed without merging**